### PR TITLE
lib/generators: add RFC42-style libconfig format

### PR DESCRIPTION
**Pull request by ckiee**
###### Description of changes

Hey again. I've split out the `libconfig` serializer from https://github.com/NixOS/nixpkgs/pull/167388 and took the time to acknowledge a few more edge-cases, as well as adding a test and auto-validator.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by infinisil** [comment]
I'm wondering, wouldn't it be better to write these generators in a generic programming language, converting from the `builtins.toJSON` output to the desired format? While this wouldn't work for `lib.generators` since it needs to return a Nix string, it should work for `pkgs.formats`, which is mostly what's used for configuration generation nowadays. The benefits would be faster evaluation, ability to reuse existing tooling, better testability.

**Comment by ckiee** [comment]
@infinisil: I think that is the way forward generally but `libconfig` doesn't really have much of an ecosystem or any CLI tools as far as I can tell so I'd like to continue with this PR as-is. I'm already 3 layers deep (https://github.com/NixOS/nixpkgs/pull/165936 -> https://github.com/NixOS/nixpkgs/pull/167388 -> [this PR] -> taking over as upstream maintainer soon most likely) and would really rather not recurse more.

File: lib/generators.nix
Comment (roberth) [inline lib/generators.nix:467]: Please don't assign meaning without context. Such assumptions are likely to lead to bugs.

```
"12A" -> "12A"
"12B" -> "12B"
...
"12L" -> 12
```

This is not how Nix works. Strings are strings. If you want a number, you write a number, or call a function that returns one.

Instead, you could represent these special with special wrappers instead `mkOctal`, etc, but do you really need multiple ways to write the same number?

File: lib/generators.nix
Comment (roberth) [inline lib/generators.nix:487]: These should be escaped and/or validated, but that would be better to do in a proper program as infinisil suggests.

**Review by roberth — CHANGES_REQUESTED** [review]
I agree with infinisil, and the semantics are fragile; see comment.

File: lib/generators.nix
Comment (ckiee) [reply lib/generators.nix:467]: > but do you really need multiple ways to write the same number?

Spec says there are a few scalar types and @anund previously said:
> [~](https://github.com/NixOS/nixpkgs/pull/167388#discussion_r933727593) This [couldn't] currently generate valid config for the example. :(

Other then that `mk{Octal|…}` seems reasonable but annoying, _future note:_ make it return just a simple `{ _type = …; … }` attrset.

I'll get to this Eventually.

**Comment by anund** [comment]
The type system for libconfig includes types not expressible directly in nix (or json). [Groups](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Groups) and [Arrays](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Arrays) can be unambiguously selected given a nix list. It's unclear if there are consumers of libconfig files that would need to care about a Group vs an Array. Expressible numbers include `0x123`, `-1`, `+1.0`, or `5L`. Octal and positive/negative I don't think are expressible. Long values might be? [Include directives](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Include-Directives) are also awkward to encode. 

For logiops in particular the usually documented config style includes references to octal number values. Ideally experimenting with packages in Nixos doesn't also come with experimenting with translating config examples. 

With that context I have trouble interpreting the intention of [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). Something like `toLibConfig` would require documenting the above mismatch everywhere it's used and generally adding `.file` inputs as an escape hatch. That sort of approach generates documentation issues. Is there a good way to handle this in Nixpkgs? Examples where nix config and file inputs are mixed perhaps?

File: lib/generators.nix
Comment (anund) [reply lib/generators.nix:467]: Playing with https://github.com/PixlOne/logiops/blob/master/logid.example.cfg as an example.

```
devices = [
  {
    #<snip> out direct translations
    buttons = [
      {
        cid: {toLibConfig.mkOctal: "0xc3"};
        action =
          {
            type: "Gestures";
            gestures: [
              {
                direction: "Up";
                mode: "OnRelease";
                action =
                {
                  type: "Keypress";
                  keys: {toLibConfig.mkArray: ["KEY_UP"]};
                };
              }
            ];
          };
      }
    ];
  }
];
```
Is that the sort of wrappers you have in mind @roberth?

**Comment by ckiee** [comment]
> The type system for libconfig includes types not expressible directly in nix (or json). [Lists](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Lists) and [Arrays](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Arrays) can't be unambiguously selected given a nix list. (a List is close to a nix list, an Array must contain only scalars of the same type). It's unclear if there are consumers of libconfig files that would need to care about a Group vs Array vs List. 

In this PR I poked your patch [a bit more](https://github.com/NixOS/nixpkgs/pull/208747/files#diff-56d5f4426e78a9b4d1d89abbfe85a6196ea411b7eaddfc51a677e1e4e55c1e91R421) to get the behavior closer to what we want. (though I don't remember the specifics very confidently ATM) and perfection is the enemy of good. (no `toLibconfig` currently!)

> Expressible numbers include `0x123`, `-1`, `+1.0`, or `5L`. Octal and positive/negative I don't think are expressible. Long values might be? 

deferring on @roberth's thread →

> [Include directives](https://hyperrealm.github.io/libconfig/libconfig_manual.html#Include-Directives) are also awkward to encode.

Yes and although I'm not sure if these should block this PR I think `"@include" = ` keys as (also similarly discussed in [the previous thread (@#167388)](https://github.com/NixOS/nixpkgs/pull/167388#discussion_r933731656)) are at least partially no go due to the awkward merging config value merging: Two includes? Maybe. `"foo.list"` and `"foo.array"`? Definitely not.

> 
> For logiops in particular the usually documented config style includes references to octal number values. Ideally experimenting with packages in [NixOS] doesn't also come with experimenting with translating config examples.
>

This is kind of a wider "found problem" with RFC 42 and I have ran into it before but it's usually been resolved quickly by opening the docs and looking at the example or defaults. (see [@#167388](https://github.com/NixOS/nixpkgs/pull/167388/files#diff-4901e9f55360e41708b8b543b7f76794f951ce45d443d1fbf4b405603f5f8a21R23))

> With that context I have trouble interpreting the intention of [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). Something like `toLibConfig` would require documenting the above mismatch everywhere it's used and generally adding `.file` inputs as an escape hatch. That sort of approach generates documentation issues. Is there a good way to handle this in Nixpkgs? Examples where nix config and file inputs are mixed perhaps?

RFC42 is already [widely deployed](https://github.com/NixOS/nixpkgs/pulls?q=rfc42) (search has [51](https://github.com/NixOS/nixpkgs/pulls?q=rfc42+is%3Amerged) merged results) and with [all](https://github.com/ckiee/nixfiles/tree/3f41c82f2776f9720e52e439bf65a9a06029916f/modules/) of [the things](https://github.com/ckiee/nixfiles/tree/3f41c82f2776f9720e52e439bf65a9a06029916f/modules/services/) I've tried so far this hasn't been that much of a problem? `libconfig` isn't too complicated a format considering #192285 `akkoma: init at 3.5.0` was recently merged using [this `elixrConf` voodoo](https://github.com/ckiee/nixpkgs/blob/fb8147aea4831984da70c8e26b2097022f9bee24/pkgs/pkgs-lib/formats.nix#L223).

I'm pretty sure neither of us are calibrated *just* right, but after all of the nixpkgs NixOS modules I've consumed I have yet to touch `environment.etc` in my own config, which is always there along with `mkForce`.

File: pkgs/pkgs-lib/formats.nix
Comment (anund) [inline pkgs/pkgs-lib/formats.nix:440]: Neat Idea. Did you consider either `<<-EOF` or a few more lines to take the file path as an input parameter with the C in a separate file?

**Comment by anund** [comment]
You are right, this work is in the land of compromise. I'm no expert in what's being done in the scope of nix options. Consider the above a mixture of documentation on the mismatch between nix and libconfig and an open ended request for a pattern, any pattern, to use in cases where nix types are insufficient for the target config schema. (I favour a `.file` option with my limited experience)

One of the least enjoyable loops in nix happens when grabbing a new package with a nix options example of `{}`. This typically involves needing to; learn a new config format, learn a new translation scheme, and iterate through a cycle of which step is wrong? the nix -> the generated config -> the program + config. That's a lot of the motivation for my original requests for examples and attempts to limit the need to mentally map things like 'the source docs use octal but decimal will also work'.

File: pkgs/pkgs-lib/formats.nix
Comment (ckiee) [reply pkgs/pkgs-lib/formats.nix:440]: I think I wanted to do something like that before it got lost in organizational hell, done now. Thanks :)

(Also, confirmed validator doesn't rebuild every time now)

File: lib/generators.nix
Comment (roberth) [reply lib/generators.nix:467]: If https://github.com/NixOS/nix/issues/7578 is implemented, the hex and octal numbers won't need any sort of escaping or wrapping. That is, unless the program parsing the config makes decisions based on the syntax used, which would be decimal regardless of what base the module user used to configured the number.

I don't think we have many examples of wrappers yet, but in the module system we generally use a representation that's an attrset with a `_type` attribute. For example:

```nix
  mkOverride = priority: content:
    { _type = "override";
      inherit priority content;
    };
```

I don't think we have any examples of formats using this approach. (I kind of expected at least something RFC42-like to use this, so sorry for my brevity earlier)
We could add a general `mkUnescape` to `lib`, but otherwise I don't know exactly where to put it. Perhaps we could have `types.foo.type` and also `types.foo.lib`?
The meaning of this `mkUnescape` is that it would take a string and concatenate it directly into the config file without any processing (except perhaps re-indenting, for some formats). This is in contrast to regular strings, which should be escaped to become strings in the target config language. This means that the string is reinterpreted according to the target configuration language syntax, and this would allow a user to specify any syntax, assuming it's going to be valid in the end. This includes such things as hex: `lib.mkUnescape "0xc3"` would do the job.

cc @infinisil

Comment (roberth) [reply lib/generators.nix:467]: That's not to say that a `formats.libconfig.mkOctal` wouldn't be nice to have.

Comment (anund) [reply lib/generators.nix:467]: A quick survey: octal is 0o[0-7]+ in python/haskell, 0[0-7]+ in java/c/golang/libconfig. This might be a format that requires a wrapper. (or a pattern of read as octal write as decimal?)

**Comment by ckiee** [comment]
Proceeded by #246115.

**Pull request closed without merging**
